### PR TITLE
Add support for Arduino Uno and 2560

### PR DESCRIPTION
**Pull request by mutantbob**
This patch includes changes I needed to build my Arduino Uno sketches.  I incorporated the work from https://github.com/japaric/heapless/pull/264 to reduce the chance of conflicts.

Now that atomic-polyfill 0.1.8 supports AVR, things are a bit easier.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -39,6 +39,13 @@ atomic-polyfill = { version = "0.1.4", optional = true }
 [target.riscv32imc-unknown-none-elf.dependencies]
 atomic-polyfill = { version = "0.1.4", optional = true }
 
+[target.avr-atmega328p.dependencies]
```
Comment (japaric) [inline Cargo.toml:42]: these targets don't appear in the output of `rustc --print target-list`. the target should be `avr-unknown-gnu-atmega328`

File: CHANGELOG.md [hunk 2]
```diff
@@ -12,6 +12,9 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
 * Fixed `pool` example in docstring.
 
 ### Added
+- Added support for AVR architecture.
+
+- Added support for AVR architecture
```
Comment (japaric) [inline CHANGELOG.md:17]: this line has been duplicated

File: build.rs [hunk 3]
```diff
@@ -43,6 +43,8 @@ fn main() -> Result<(), Box<dyn Error>> {
     match &target[..] {
         "avr-unknown-gnu-atmega328"
         | "msp430-none-elf"
+        | "avr-atmega328p"
```
Comment (japaric) [inline build.rs:46]: should be `avr-unknown-gnu-atmega328`

File: build.rs [hunk 4]
```diff
@@ -55,7 +57,10 @@ fn main() -> Result<(), Box<dyn Error>> {
     // Let the code know if it should use atomic-polyfill or not, and what aspects
     // of polyfill it requires
     match &target[..] {
-        "riscv32i-unknown-none-elf" | "riscv32imc-unknown-none-elf" => {
+        "riscv32i-unknown-none-elf"
+        | "riscv32imc-unknown-none-elf"
+        | "avr-atmega328p"
```
Comment (japaric) [inline build.rs:62]: should be `avr-unknown-gnu-atmega328`

**Review by japaric — CHANGES_REQUESTED** [review]

File: Cargo.toml (hunk 1 above)
Comment (mutantbob) [reply Cargo.toml:39]: If I make the change you suggest, I get the following error:

```
error[E0432]: unresolved import `atomic_polyfill`
  --> /home/thoth/src/heapless/src/pool/singleton/arc.rs:85:5
   |
85 | use atomic_polyfill::{AtomicUsize, Ordering};
   |     ^^^^^^^^^^^^^^^ use of undeclared crate or module `atomic_polyfill`

error[E0432]: unresolved import `atomic_polyfill`
 --> /home/thoth/src/heapless/src/pool/llsc.rs:7:5
  |
7 | use atomic_polyfill::{AtomicPtr, Ordering};
  |     ^^^^^^^^^^^^^^^ use of undeclared crate or module `atomic_polyfill`

error[E0433]: failed to resolve: use of undeclared crate or module `atomic_polyfill`
  --> /home/thoth/src/heapless/src/mpmc.rs:97:25
   |
97 | type AtomicTargetSize = atomic_polyfill::AtomicU8;
   |                         ^^^^^^^^^^^^^^^ use of undeclared crate or module `atomic_polyfill`

error[E0433]: failed to resolve: use of undeclared crate or module `atomic_polyfill`
   --> /home/thoth/src/heapless/src/mpmc.rs:102:17
    |
102 | type Ordering = atomic_polyfill::Ordering;
    |                 ^^^^^^^^^^^^^^^ use of undeclared crate or module `atomic_polyfill`

error[E0392]: parameter `T` is never used
   --> /home/thoth/src/heapless/src/pool/mod.rs:263:17
    |
263 | pub struct Pool<T> {
    |                 ^ unused parameter
    |
    = help: consider removing `T`, referring to it in a field, or using a marker such as `PhantomData`
    = help: if you intended `T` to be a const parameter, use `const T: usize` instead
```

File: build.rs (hunk 3 above)
Comment (mutantbob) [reply build.rs:48]: I have discovered that I can just remove these match clauses, so I am.

File: build.rs (hunk 4 above)
Comment (mutantbob) [reply build.rs:64]: If I perform this modification it malfunctions like this:

```
  --> /home/thoth/src/heapless/src/pool/singleton/arc.rs:88:26
   |
88 | use core::sync::atomic::{AtomicUsize, Ordering};
   |                          ^^^^^^^^^^^ no `AtomicUsize` in `sync::atomic`

error[E0432]: unresolved import `core::sync::atomic::AtomicPtr`
  --> /home/thoth/src/heapless/src/pool/llsc.rs:10:26
   |
10 | use core::sync::atomic::{AtomicPtr, Ordering};
   |                          ^^^^^^^^^
   |                          |
   |                          no `AtomicPtr` in `sync::atomic`
   |                          help: a similar name exists in the module: `AtomicI8`

error[E0432]: unresolved import `core::sync::atomic::AtomicUsize`
  --> /home/thoth/src/heapless/src/spsc.rs:92:26
   |
92 | use core::sync::atomic::{AtomicUsize, Ordering};
   |                          ^^^^^^^^^^^ no `AtomicUsize` in `sync::atomic`

error[E0392]: parameter `T` is never used
   --> /home/thoth/src/heapless/src/pool/mod.rs:263:17
    |
263 | pub struct Pool<T> {
    |                 ^ unused parameter
    |
    = help: consider removing `T`, referring to it in a field, or using a marker such as `PhantomData`
    = help: if you intended `T` to be a const parameter, use `const T: usize` instead
```

**Comment by mutantbob** [comment]
Several of your requested changes seem to revolve around the assumption that the target will be `avr-unknown-gnu-atmega328p` or similar.  This just is not the case when compiling for AVR.

My current hypothesis is that the target string comes from the name of the `avr-specs/*.json` files at https://github.com/Rahix/avr-hal-template/tree/main/avr-specs .

Can you point at a Rust application that targets AVR and does have `target=avr-unknown-gnu-atmega328p` ?

**Comment by japaric** [comment]
> Several of your requested changes seem to revolve around the assumption that the target will be avr-unknown-gnu-atmega328p or similar.

`avr-unknown-gnu-atmega328` is the **only** *built-in* AVR target as of Rust 1.61.0. see below

``` console
$ rustc --print target-list | rg avr
avr-unknown-gnu-atmega328
```

> My current hypothesis is that the target string comes from the name of the avr-specs/*.json files at https://github.com/Rahix/avr-hal-template/tree/main/avr-specs .

These are *custom* compilation targets. Their names can be arbitrary, e.g. a target named `arm-cortex-m4.json` could be an AVR target. It's not reasonable to expect any crate to accommodate *custom* targets by **name**. if you need to cover those custom targets via conditional compilation you should use `#[cfg(target_arch = "avr")]` or some other target specification option that they share in common. see `rustc --print cfg --target my-custom-target`

**Comment by japaric** [comment]
thanks! looks good now.

bors r+

**Comment by japaric** [comment]
release v0.7.14 includes this PR

**Pull request merged**